### PR TITLE
EDLY-3212 Added logging for exceptions to add fault tolerance in job

### PR DESCRIPTION
**Pull request by iam-mhaseeb**
**Description:** Added logging for exceptions to add fault tolerance in job.
**JIRA:** https://edlyio.atlassian.net/browse/EDLY-3212

### CODE REVIEW

**Review by burhandodhy — APPROVED** [review]

**Review by etechdevai — APPROVED** [review]
👍

**Review by taimoor-ahmed-1 — APPROVED** [review]

File: credentials/apps/catalog/management/commands/copy_catalog.py [hunk 1]
```diff
@@ -41,10 +42,28 @@ def handle(self, *args, **options):
 
             logger.info('Copying catalog data for site {}'.format(site.domain))
             client = site_config.catalog_api_client
-            Command.fetch_programs(site, client, page_size=page_size)
-            logger.info('Finished copying programs.')
-            Command.fetch_pathways(site, client, page_size=page_size)
-            logger.info('Finished copying pathways.')
+            try:
+                Command.fetch_programs(site, client, page_size=page_size)
+                logger.info('Finished copying programs.')
```
Comment (zubair-arbi) [inline credentials/apps/catalog/management/commands/copy_catalog.py:48]: Add counter for the successful sites

File: credentials/apps/catalog/management/commands/copy_catalog.py [hunk 2]
```diff
@@ -41,10 +42,28 @@ def handle(self, *args, **options):
...
+            except Exception as error:
+                success = False
+                logger.info('Failed to copy programs for site: {} with error: {}'.format(
```
Comment (zubair-arbi) [inline credentials/apps/catalog/management/commands/copy_catalog.py]: Exception log

File: credentials/apps/catalog/management/commands/copy_catalog.py [hunk 3]
```diff
@@ -41,10 +42,28 @@ def handle(self, *args, **options):
...
+                    site.domain,
+                    error
+                ))
+
+            try:
+                Command.fetch_pathways(site, client, page_size=page_size)
+                logger.info('Finished copying pathways.')
+            except Exception as error:
+                success = False
+                logger.info('Failed to copy pathways for site: {} with error: {}'.format(
```
Comment (zubair-arbi) [inline credentials/apps/catalog/management/commands/copy_catalog.py]: exception log

File: credentials/apps/catalog/management/commands/copy_catalog.py [hunk 4]
```diff
@@ -41,10 +42,28 @@ def handle(self, *args, **options):
...
+                    site.domain,
+                    error
+                ))
+
+        if not success:
+            raise CommandError('One or more of the sites loaders above failed.')
```
Comment (zubair-arbi) [inline credentials/apps/catalog/management/commands/copy_catalog.py]: `Programs / Pathways copying for 3 out of 6 sites have failed. Please update the configuration.`

**Review by zubair-arbi — CHANGES_REQUESTED** [review]
I have added my feedback.

File: credentials/apps/catalog/management/commands/copy_catalog.py (hunk 1 above)
Comment (iam-mhaseeb) [reply credentials/apps/catalog/management/commands/copy_catalog.py:48]: Done

File: credentials/apps/catalog/management/commands/copy_catalog.py (hunk 3 above)
Comment (iam-mhaseeb) [reply credentials/apps/catalog/management/commands/copy_catalog.py]: Done

File: credentials/apps/catalog/management/commands/copy_catalog.py (hunk 2 above)
Comment (iam-mhaseeb) [reply credentials/apps/catalog/management/commands/copy_catalog.py]: Done

File: credentials/apps/catalog/management/commands/copy_catalog.py (hunk 4 above)
Comment (iam-mhaseeb) [reply credentials/apps/catalog/management/commands/copy_catalog.py]: Done

File: credentials/apps/catalog/management/commands/copy_catalog.py [hunk 5]
```diff
@@ -41,10 +43,30 @@ def handle(self, *args, **options):
...
+            except Exception as error:
+                success = False
+                logger.exception('Failed to copy pathways for site: {}.'.format(site.domain))
+
+            if not success:
+                failure_count += 1
+
+        if failure_count:
+            raise CommandError(
+                'Programs / Pathways copying for {} out of {} sites have failed. Please check the configuration.'.format(
```
Comment (zubair-arbi) [inline credentials/apps/catalog/management/commands/copy_catalog.py]:
```python
        if failure_count:
            raise CommandError(
                'Programs / Pathways copying for {failure_for_sites} out of {total_sites} sites have failed. '
                'Please check the configuration.'.format(
                    failure_for_sites=failure_count,
                    total_sites=Site.objects.count(),
                )
            )
```

**Review by zubair-arbi — CHANGES_REQUESTED** [review]
I  small suggestion. Otherwise changes looks good now

File: credentials/apps/catalog/management/commands/copy_catalog.py (hunk 5 above)
Comment (iam-mhaseeb) [reply credentials/apps/catalog/management/commands/copy_catalog.py]: Updated

**Review by zubair-arbi — APPROVED** [review]
Changes looks good now. 👍

**Pull request merged**